### ISSUE TITLE
Relative paths duplicates paths on Windows

### ISSUE THREAD

**Issue by henryiii** [coveragepy/coveragepy#1752]
**Describe the bug**

I've been working on issues with nox's test suite. We weren't correctly including runs from all three OS's, and we have just been fixing that. We seem to have a 50% chance of failure with `D:\a\nox\nox\__init__.py` showing up with "no source for file". I'm pretty sure this started showing up for the first time without this fix roughly 1-2 days ago, but the issue is more reliable if we properly make sure coverage files don't overwrite each other. We are using editable installs and relative paths, so I'm not sure why it's getting stored that way (even in the succeeding 50%). If I look at the report on Windows, every file is listed twice. If I look at the `.coverage.win32*` files, I see that every file is present twice in `files`, once as a relative path, and once as an absolute path. The other OS's are just fine, it's only Windows that is doing this.

This is with a hatchling editable install of nox, inside a nox runner. Using pytest-cov.

**To Reproduce**
How can we reproduce the problem?  Please *be specific*. Don't link to a failing CI job. Answer the questions below:
1. What version of Python are you using? _3.7-3.12_.
1. What version of coverage.py shows the problem? The output of `coverage debug sys` is helpful.  _7.4.2 and 7.2.7_
1. What versions of what packages do you have installed? The output of `pip freeze` is helpful. _all latest supported_ 
1. What code shows the problem?  Give us a specific commit of a specific repo that we can check out. If you've already worked around the problem, please provide a commit before that fix. https://github.com/wntrblm/nox/commit/3b8e0267b7c2c3631d1098f8a4d6e4f731582423 passed at one point, and now is failing. https://github.com/wntrblm/nox/pull/778/commits/1c998d114dbe31c79b7a78e4010fd2a2a48008f7 is probably better, though, since that's with the OS naming fix.
1. What commands should we run to reproduce the problem? *Be specific*. Include everything, even `git clone`, `pip install`, and so on. Explain like we're five! _I've only done it in CI, since I don't have easy access to a Windows machine right now, but it should be possible to run `pipx run `nox -s 'tests(python="3.12", tox_version="latest")'` (AFAICT it happens on all of them).

**Expected behavior**
There should not be two copies of files on Windows. `SELECT * FROM file;` reports:

```
1|D:\a\nox\nox\nox\__init__.py
2|D:\a\nox\nox\nox\_options.py
3|D:\a\nox\nox\nox\_option_set.py
4|D:\a\nox\nox\nox\tasks.py
5|D:\a\nox\nox\nox\registry.py
6|D:\a\nox\nox\nox\_decorators.py
7|D:\a\nox\nox\nox\_typing.py
8|D:\a\nox\nox\nox\_version.py
9|D:\a\nox\nox\nox\logger.py
10|D:\a\nox\nox\nox\manifest.py
11|D:\a\nox\nox\nox\sessions.py
12|D:\a\nox\nox\nox\command.py
13|D:\a\nox\nox\nox\popen.py
14|D:\a\nox\nox\nox\virtualenv.py
15|D:\a\nox\nox\nox\_parametrize.py
16|D:\a\nox\nox\nox\tox_to_nox.py
17|D:\a\nox\nox\nox\workflow.py
18|D:\a\nox\nox\nox\__main__.py
19|nox\__init__.py
20|nox\_options.py
21|nox\_option_set.py
22|nox\tasks.py
23|nox\registry.py
24|nox\_decorators.py
25|nox\_version.py
26|nox\logger.py
27|nox\manifest.py
28|nox\sessions.py
29|nox\command.py
30|nox\popen.py
31|nox\virtualenv.py
32|nox\_parametrize.py
33|nox\__main__.py
34|nox\workflow.py
35|nox\tox_to_nox.py
```

And the coverage report looks like this:

```
---------- coverage: platform win32, python 3.8.10-final-0 -----------
Name                  Stmts   Miss Branch BrPart  Cover
-------------------------------------------------------
nox\__init__.py           8      0      0      0   100%
nox\__init__.py           8      0      0      0   100%
nox\__main__.py          20      0      4      0   100%
nox\__main__.py          20      0      4      0   100%
nox\_decorators.py       52      0      6      0   100%
nox\_decorators.py       52      0      6      0   100%
nox\_option_set.py      108      0     38      0   100%
nox\_option_set.py      108      0     38      0   100%
nox\_options.py          78      0     32      0   100%
nox\_options.py          78      0     32      0   100%
nox\_parametrize.py      75      0     36      0   100%
nox\_parametrize.py      75      0     36      0   100%
nox\_typing.py            4      4      0      0     0%
nox\_version.py          42      1     14      1    96%
nox\_version.py          42      1     14      1    96%
nox\command.py           68      1     34      2    97%
nox\command.py           68      1     34      2    97%
nox\logger.py            47      0     16      1    98%
nox\logger.py            47      0     16      1    98%
nox\manifest.py         150      0     84      0   100%
nox\manifest.py         150      0     84      0   100%
nox\popen.py             38      7      8      1    83%
nox\popen.py             38      7      8      1    83%
nox\registry.py          23      0      6      0   100%
nox\registry.py          23      0      6      0   100%
nox\sessions.py         303      0     96      0   100%
nox\sessions.py         303      0     96      0   100%
nox\tasks.py            152      0     56      0   100%
nox\tasks.py            152      0     56      0   100%
nox\tox_to_nox.py        64     30     28      2    43%
nox\tox_to_nox.py        64     30     28      2    43%
nox\virtualenv.py       211      3     70      5    97%
nox\virtualenv.py       211      3     70      5    97%
nox\workflow.py          18      1      6      1    92%
nox\workflow.py          18      1      6      1    92%
-------------------------------------------------------
TOTAL                  2918     90   1068     26    96%
```

Linux and macOS jobs do not show duplicate copies.

**Additional context**

~~I think I have finally been able to work around it by avoiding relative paths and using `source_pgks` and avoiding the editable install~~ (I get an issue with macOS paths if I include it, though the duplicated paths issue is still fine). I'm pretty sure that any time I specify relative paths, this is going to be duplicated, and if I don't, it's fine. I've been trying a few dozen things, across several PRs if any of the other combinations are useful.

I think I can work around it by post-processing the data files to remove the `D:\` files (I also swap `\` for `/` - if I do that locally, I can view the files without the duplicate lines).

This is my working workaround in https://github.com/wntrblm/nox/pull/778:

```python
if sys.platform.startswith("win"):
    with contextlib.closing(sqlite3.connect(coverage_file)) as con, con:
        con.execute("UPDATE file SET path = REPLACE(path, '\\', '/')")
        con.execute("DELETE FROM file WHERE SUBSTR(path, 2, 1) == ':'")
```

**Comment by nedbat** [comment]
> https://github.com/wntrblm/nox/commit/1c998d114dbe31c79b7a78e4010fd2a2a48008f7 is probably better, though, since that's with the OS naming fix.

I created a branch from that commit, and am having trouble finding and examining the coverage data files: https://github.com/nedbat/nox/actions/runs/8161334625   Can you help me understand how it's supposed to be working?  Without my changes, I see that the "Save coverage report" step is finding files to upload, so I'm confused.

**Comment by nedbat** [comment]
I did some experiments in an Azure virtual machine.  It looks to me like the data attributed to `nox\something.py` is collected in the main process, and the data attributed to `D:\a\nox\nox\nox\something.py` is being collected in subprocesses.  Is there something unusual about how nox or its test suite manages subprocesses?

**Comment by nedbat** [comment]
I'm also wondering if you can run this scenario with coverage.py directly instead of using pytest-cov, just to remove one factor.

**Comment by henryiii** [comment]
I can try. My baby daughter came home from the hospital yesterday so won't have much time for a bit, but I'll try without pytest-cov at some point in the near(ish) future!

**Comment by BruDriguezz** [comment]
.

**Comment by nedbat** [comment]
I have found the heart of the problem:

When combining data files, paths get remapped.  With relative_files=true, a mapping rule is auto-generated when a relative path name is seen.  That rule can also apply to absolute paths.  Here's an excerpt from the debug log (with `debug=pathmap`):
```
Aliases (relative=True):
Generating rule: '*/nox' -> 'nox\\' using regex '^(.*[\\\\/])?nox[\\\\/]'
Matched path 'nox\\__init__.py' to rule '*/nox' -> 'nox\\', producing 'nox\\__init__.py'
... many lines skipped
Matched path 'C:\\Users\\nedbat\\nox\\nox\\__init__.py' to rule '*/nox' -> 'nox\\', producing 'nox\\__init__.py'
```

If the first file combined has absolute paths, we see this instead:
```
Aliases (relative=True):
No rules match, path 'C:\\Users\\nedbat\\nox\\nox\\__init__.py' is unchanged
```

The bug was happening only sometimes because sometimes the first file combined at absolute paths.  Then there hadn;t been a remapping rule generated, so the absolute paths remained unchanged.  If the first file to be combined had relative paths, then the rule would be generated, and when the absolute paths were encountered, they would be remapped.

The order of the files being combined depended on the file name, which includes a process id, that's the source of the randomness.  If the first file (ordered by process id) had absolute paths, then there were duplicates in the result.  if the first file had relative paths, then there weren't duplicates because the mapping rule was autogenerated and absolute paths were mapped to relative.

Why is this only happening on Windows?  I'm not sure why, but there are more sub-processes spawned on Windows than on Mac.  The data files with the absolute path names are created by those extra subprocesses.

**Comment by nedbat** [comment]
@BruDriguezz let me know if you need more info

**Comment by nedbat** [comment]
This is fixed in 1b19799edfbfc65dea254e8fcaa0dea23e9709f9.

**Comment by nedbat** [comment]
This is now released as part of [coverage 7.4.4](https://pypi.org/project/coverage/7.4.4).

**Comment by jab** [comment]
Just saw the latest release notes and found my way here, since it reminded me of an issue I opened a few months ago (#1674). I am no longer affected by that issue (new job), but figured I'd link to it from here in case [the fix for this](https://github.com/nedbat/coveragepy/commit/1b19799edfbfc65dea254e8fcaa0dea23e9709f9) fixes #1674 too.
